### PR TITLE
[flutter_tools] Update background isolates when performing hot reload/restart

### LINKED ISSUE #47646: Background isolates aren't killed when hot restarting
`flutter run` the program below. The output will say "Main thread main" and "Isolate thread background" every three seconds. Hit `R` in the console and now you'll have two background thread messages for every main thread message.

```dart
import 'dart:async';
import 'dart:isolate';

import 'package:flutter/material.dart';

void main() {
  Isolate.spawn<void>(background, null, debugName: 'background');
  Timer.periodic(const Duration(seconds: 3), (Timer timer) {
    new TestMain();
  });
}

void background(void message) {
  Timer.periodic(const Duration(seconds: 3), (Timer timer) {
    new TestIsolate();
  });
}

class TestMain {
  TestMain() {
    debugPrint('Main thread ${Isolate.current.debugName}');
  }
}

class TestIsolate {
  TestIsolate() {
    debugPrint('Isolate thread ${Isolate.current.debugName}');
  }
}
```

cc @bkonyi @jonahwilliams @jason-simmons 

See also https://github.com/flutter/flutter/issues/20743 (hot reload doesn't update background isolates).

### LINKED ISSUE #48091: Hot reload doesn't update static methods used in other isolates
Steps to reproduce:

1. `flutter create hot_reload_isolate_test`
2. Replace the `hot_reload_isolate_test/lib/main.dart` file with the following:
```dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: Material(
        child: Builder(
          builder: (BuildContext context) => Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              RaisedButton(
                child: Text('FROM SAME ISOLATE'),
                onPressed: () async {
                  await showMessage(context, _test(null));
                },
              ),
              RaisedButton(
                child: Text('FROM OTHER ISOLATE'),
                onPressed: () async {
                  await showMessage(context, await compute<void, String>(_test, null));
                },
              ),
            ],
          ),
        ),
      ),
    );
  }
}

Future<void> showMessage(BuildContext context, String message) {
  return showDialog(
    context: context,
    builder: (BuildContext context) => AlertDialog(
      content: Text(message),
    ),
  );
}

String _test(void argument) {
  return 'Hello';
}
```
3. Go into the `hot_reload_isolate_test` directory and `flutter run`.
4. You should get an app with two buttons, both of which cause a dialog to appear saying "HELLO".
5. Edit the file above, changing `Hello` on the penultimate line to `Goodbye`.
6. Hot reload and push the buttons again.

Expected results: Both buttons should now say "Goodbye".

Actual results: the button that spawns a different isolate still says "Hello".

cc @bkonyi

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

When performing a hot restart, collect isolates without an attached flutter view and send a kill signal. These must have been spawned by running `main`, so restarting without removing them leads to isolate duplication.

This works provided that there are no cases where these isolates should not be removed...

When performing a hot reload, ensure that we send a reloadSources command to every isolate and not just uiIsolates.

Fixes https://github.com/flutter/flutter/issues/47646
Fixes https://github.com/flutter/flutter/issues/48091 (but only for existing isolates).

There are two edge cases here:
  1. plugins that communicate with isolates launched via the main isolate
  2. background isolates

This change might shutoff isolates prematurely in 1), however I would argue this isn't a big deal since today these plugins would not function under a hot restart anyway (duplicates spawned).

For 2, since we do not kill isolates with a FlutterView attached (which includes android alarm manager) there is no difference in functionality, though the isolate itself still can't be restarted.

### CODE REVIEW

File: packages/flutter_tools/lib/src/run_hot.dart [hunk 1]
```diff
@@ -788,13 +792,6 @@ class HotRunner extends ResidentRunner {
       }
     }
 
-    // The initial launch is from a script snapshot. When we reload from source
-    // on top of a script snapshot, the first reload will be a worst case reload
-    // because all of the sources will end up being dirty (library paths will
-    // change from host path to a device path). Subsequent reloads will
-    // not be affected, so we resume reporting reload times on the second
-    // reload.
-    bool shouldReportReloadTime = !_runningFromSnapshot;
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/run_hot.dart]: Removed comment since this is no longer true

**Comment by jonahwilliams** [comment]
Need to update mocks in the resident runner test, otherwise this works as is

**Comment by bkonyi** [comment]
This approach generally LGTM. Thanks for fixing this!

**Comment by jonahwilliams** [comment]
PTAL

**Comment by bkonyi** [comment]
> For 2, since we do not kill isolates with a FlutterView attached (which includes android alarm manager) there is no difference in functionality, though the isolate itself still can't be restarted.

The alarm manager plugin shouldn't have a view attached, unless a `FlutterEngine` instance does by default?

**Comment by jonahwilliams** [comment]
I think flutter view = FlutterEngine instance, which is what I see when I run an app using android alarm manager:

```
I/AlarmService(21540): AlarmService started!                            
Syncing files to device Pixel 4...                                 486ms

Flutter run key commands.
r Hot reload. 🔥🔥🔥
R Hot restart.
h Repeat this help message.
d Detach (terminate "flutter run" but leave application running).
c Clear the screen
q Quit (terminate the application on the device).
An Observatory debugger and profiler on Pixel 4 is available at: http://127.0.0.1:64837/-j1Zy1FiVrk=/
I/flutter (21540): >>> GET /api/v2/event (body length: null)
I/flutter (21540): >>> GET /api/v2/announcements (body length: null)

Connected views:
  main.dart$main-639014536911191 (isolates/639014536911191)
  main.dart$main-4274356730205947 (isolates/4274356730205947)
```

**Comment by bkonyi** [comment]
Ah gotcha. In that case we'll want to figure out how we're going to deal with isolates held via a `FlutterEngine`. Otherwise, this change LGTM for regularly spawned isolates.

**Review by zanderso — APPROVED** [review]
lgtm w/ more comments

File: packages/flutter_tools/lib/src/run_hot.dart [hunk 2]
```diff
@@ -519,15 +519,18 @@ class HotRunner extends ResidentRunner {
         device.generator.accept();
       }
     }
-    // Check if the isolate is paused and resume it.
-    final List<Future<void>> futures = <Future<void>>[];
+    // Check if the isolate is paused and resume it. Tag isolates
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/run_hot.dart]: I added more documentation inline with the isolate killing to explain it for now. I agree we should probably have the entire process outlined (for both hot reload and hot restart) but I'm quite sure yet of the best place to document it. The prior comments here fell out of sync pretty quickly

**Pull request merged**